### PR TITLE
[Doc][AIR] Improve visibility of Trainer restore and stateful callback restoration

### PR DESCRIPTION
**Pull request by justinvyu**
## Why are these changes needed?

This PR moves the FAQ section on Train experiment restoration to the DL/GBDT user guides. Plus this PR makes these code snippets tested. Secondly, this PR improves the docstrings of methods to implement for saving/restoring stateful Callbacks.

### Links for reviewers

- https://anyscale-ray--34350.com.readthedocs.build/en/34350/train/dl_guide.html#fault-tolerance
- https://anyscale-ray--34350.com.readthedocs.build/en/34350/train/gbdt.html#saving-and-loading-xgboost-and-lightgbm-checkpoints
- https://anyscale-ray--34350.com.readthedocs.build/en/34350/train/config_guide.html

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/train/gbdt.rst [hunk 1]
```diff
@@ -33,8 +33,34 @@ training parameters are passed as the ``params`` dictionary.
...
+
+.. literalinclude:: doc_code/key_concepts.py
+    :language: python
+    :start-after: __checkpoint_config_ckpt_freq_start__
+    :end-before: __checkpoint_config_ckpt_freq_end__
+
+.. tip::
+
+    Once checkpointing is enabled, you can follow :ref:`this guide <train-fault-tolerance>`
+    to enable fault tolerance.
```
Comment (matthewdeng) [inline doc/source/train/gbdt.rst:56]: Probably don't want to link this since this is in the Deep Learning guide.

**Review by matthewdeng — APPROVED** [review]
Very nice!

File: doc/source/train/config_guide.rst [hunk 2]
```diff
@@ -30,13 +33,20 @@ resources per worker.
```
[the remark this answers is not in the document]
Comment (justinvyu) [reply doc/source/train/config_guide.rst:33]: Oh, that one is for the previous section on the `ScalingConfig`. I put the API ref links at the bottom of each section. Can move it if that's confusing.

File: doc/source/train/gbdt.rst (hunk 1 above)
Comment (justinvyu) [reply doc/source/train/gbdt.rst:56]: Yeah, I was trying to share some content between the two. I linked it to a section that doesn't really depend on DL trainers. I think it's okay for now, and we can rethink the docs/user guide structure as part of the docs side of the layering project. wdyt?

**Pull request merged**